src/charm.py

```diff
@@ -191,185 +190,101 @@ def unit_peer_data(self) -> Dict:
-            return
-        if scope == UNIT_SCOPE:
-            self.unit_peer_data.update({key: value})
+            secret_key = self._translate_field_to_secret_key(key)
+            label = generate_secret_label(self, scope)
+            secret = self.secrets.get(label)
+            if not secret:
+                self.secrets.add(label, {secret_key: value}, scope)
+            else:
+                content = secret.get_content()
+                content.update({secret_key: value})
+                secret.set_content(content)
+            return label
         else:
-            self.app_peer_data.update({key: value})
-
-    def _juju_secret_remove(self, scope: str, key: str) -> None:
-        """Remove a Juju 3.x secret."""
-        self._juju_secrets_get(scope)
+            self._peer_data(scope).update({key: value})
 
-        key = SECRET_KEY_OVERRIDES.get(key, key)
+    def remove_secret(self, scope: Scopes, key: str) -> None:
+        """Removing a secret."""
+        if scope not in get_args(Scopes):
+            raise RuntimeError("Unknown secret scope.")
 
-        secret = self.secrets[scope].get(SECRET_LABEL)
-        if not secret:
-            logging.error(f"Secret {scope}:{key} wasn't deleted: no secrets are available")
-            return
+        if JujuVersion.from_environ().has_secrets:
+            secret_key = self._translate_field_to_secret_key(key)
+            label = generate_secret_label(self, scope)
+            secret = self.secrets.get(label)
 
-        secret_cache = self.secrets[scope].get(SECRET_CACHE_LABEL)
-        if not secret_cache or key not in secret_cache:
-            logging.error(f"No secret {scope}:{key}")
-            return
+            if not secret:
+                return
 
-        secret_cache[key] = SECRET_DELETED_LABEL
-        secret.set_content(secret_cache)
-        logging.debug(f"Secret {scope}:{key}")
+            content = secret.get_content()
 
-        # TODO change upgrade to switch to secrets once minor version upgrades is done
-        if scope == UNIT_SCOPE:
-            peer_data = self.unit_peer_data
-        else:
-            peer_data = self.app_peer_data
-        if key in peer_data:
-            del peer_data[key]
-
-    def remove_secret(self, scope: str, key: str) -> None:
-        """Removing a secret."""
-        if scope not in [APP_SCOPE, UNIT_SCOPE]:
-            raise RuntimeError("Unknown secret scope.")
+            if not content.get(secret_key) or content[secret_key] == SECRET_DELETED_LABEL:
+                logger.error(f"Non-existing secret {scope}:{key} was attempted to be removed.")
+                return
 
-        juju_version = JujuVersion.from_environ()
-        if juju_version.has_secrets:
-            return self._juju_secret_remove(scope, key)
-        if scope == UNIT_SCOPE:
-            del self.unit_peer_data[key]
+            content[secret_key] = SECRET_DELETED_LABEL
+            secret.set_content(content)
         else:
```

Comment:
We should always pop here, secret might have been originally set in the databag and never moved to secrets.